src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
+        internal readonly ref struct DiyFp
         {
-            public const int SignificandLength = 64;
+            public const int SignificandSize = 64;
 
-            // Extended significand.
-            // IEEE 754 double-precision numbers only require 53 bits significand.
-            // However, in Grisu3 we need additional 11 bits so we declare _f as ulong.
-            // Please note _f does not include sign bit.
-            private ulong _f;
+            public readonly ulong f;
+            public readonly int e;
 
-            // Extended exponent.
-            // IEEE 754 double-precision numbers only require 11 bits exponent.
-            // However, in Grisu3 we need extra space to avoid overflow so we declare _e as int.
-            // Please note _e is a biased exponent if the DiyFp instance is generated by GenerateNormalizedDiyFp().
-            private int _e;
-
-            public DiyFp(ulong f, int e)
+            public DiyFp(double value)
             {
-                _f = f;
-                _e = e;
+                Debug.Assert(double.IsFinite(value));
+                Debug.Assert(value > 0.0);
+                f = ExtractFractionAndBiasedExponent(value, out e);
             }
 
-            public ulong f
+            public DiyFp(float value)
             {
-                get
-                {
-                    return _f;
-                }
+                Debug.Assert(float.IsFinite(value));
+                Debug.Assert(value > 0.0f);
+                f = ExtractFractionAndBiasedExponent(value, out e);
+            }
 
-                set
-                {
-                    _f = value;
-                }
+            // Computes the two boundaries of value.
+            //
+            // The bigger boundary (mPlus) is normalized.
+            // The lower boundary has the same exponent as mPlus.
+            //
+            // Precondition:
+            //  The value encoded by value must be greater than 0.
+            public DiyFp(double value, out DiyFp mMinus, out DiyFp mPlus) : this(value)
+            {
+                GetBoundaries(52, out mMinus, out mPlus);
             }
 
-            public int e
+            // Computes the two boundaries of value.
+            //
+            // The bigger boundary (mPlus) is normalized.
+            // The lower boundary has the same exponent as mPlus.
+            //
+            // Precondition:
+            //  The value encoded by value must be greater than 0.
+            public DiyFp(float value, out DiyFp mMinus, out DiyFp mPlus) : this(value)
             {
-                get
-                {
-                    return _e;
-                }
+                GetBoundaries(23, out mMinus, out mPlus);
+            }
 
-                set
-                {
-                    _e = value;
-                }
+            public DiyFp(ulong significand, int exponent)
+            {
+                f = significand;
```

Comment:
I think this constructor shows that `f` and `e` are probably bad names for the field... typically the name of the constructor parameters match the name of the field/property they set.

[https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/constructor](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/constructor)

✓ DO use the same name for constructor parameters and a property if the constructor parameters are used to simply set the property.

Reply:
These are used in order to match the names used in the paper and reference implementation, which makes it easier to assert that the implementation is correct.

Reply:
`f` and `e` are also the typical shortnames used for `significand` and `exponent`

Reply:
Then let's use that in the constructor parameters.

---
In reply to: [252369357](https://github.com/dotnet/coreclr/pull/22040#discussion_r252369357) [](ancestors = 252369357)

Reply:
The reference implementation uses the same convention: <https://github.com/google/double-conversion/blob/master/double-conversion/diy-fp.h#L45>

It makes it clear in the source that `f` is `significand` and that `e` is `exponent`, while allowing the consuming code to best match the math/algorithms listed in the paper.

I don't feel particularly strongly about this, and will update to just use `f` and `e` everywhere if you feel otherwise.

Reply:
Yes, using `f` and `e` everywhere makes sense to me. If everything is using that name, it makes it better than some places using one word, and another using a different word.

Reply:
Updated.